core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -0,0 +1,719 @@
+            }
+            if (this.leaderEpoch != leaderEpochVal) {
+                leaderEpoch = leaderEpochVal;
+            }
+            // Reset readOffset, so that it is set in next run of RLMTask
+            copiedOffsetOption = OptionalLong.empty();
+        }
+
+        public void convertToFollower() {
+            leaderEpoch = -1;
+        }
+
+        private void maybeUpdateReadOffset() throws RemoteStorageException {
+            if (!copiedOffsetOption.isPresent()) {
+                logger.info("Find the highest remote offset for partition: {} after becoming leader, leaderEpoch: {}", topicIdPartition, leaderEpoch);
+
+                // This is found by traversing from the latest leader epoch from leader epoch history and find the highest offset
+                // of a segment with that epoch copied into remote storage. If it can not find an entry then it checks for the
+                // previous leader epoch till it finds an entry, If there are no entries till the earliest leader epoch in leader
+                // epoch cache then it starts copying the segments from the earliest epoch entry’s offset.
+                copiedOffsetOption = OptionalLong.of(findHighestRemoteOffset(topicIdPartition));
+            }
+        }
+
+        public void copyLogSegmentsToRemote() throws InterruptedException {
+            if (isCancelled())
+                return;
+
+            try {
+                maybeUpdateReadOffset();
+                long copiedOffset = copiedOffsetOption.getAsLong();
+                Optional<UnifiedLog> maybeLog = fetchLog.apply(topicIdPartition.topicPartition());
+                if (!maybeLog.isPresent()) {
+                    return;
+                }
+
+                UnifiedLog log = maybeLog.get();
+
+                // LSO indicates the offset below are ready to be consumed (high-watermark or committed)
+                long lso = log.lastStableOffset();
+                if (lso < 0) {
+                    logger.warn("lastStableOffset for partition {} is {}, which should not be negative.", topicIdPartition, lso);
+                } else if (lso > 0 && copiedOffset < lso) {
+                    // Copy segments only till the last-stable-offset as remote storage should contain only committed/acked
+                    // messages
+                    long toOffset = lso;
+                    logger.debug("Checking for segments to copy, copiedOffset: {} and toOffset: {}", copiedOffset, toOffset);
+                    long activeSegBaseOffset = log.activeSegment().baseOffset();
+                    // log-start-offset can be ahead of the read-offset, when:
+                    // 1) log-start-offset gets incremented via delete-records API (or)
+                    // 2) enabling the remote log for the first time
+                    long fromOffset = Math.max(copiedOffset + 1, log.logStartOffset());
```

Comment:
Maybe we should log a message (warning?) if `copiedOffset + 1` < `log.logStartOffset()` in which case there may be a gap in the remote log? What if another replica (non-leader) still holds the data we are missing here?